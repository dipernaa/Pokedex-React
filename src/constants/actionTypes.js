import { buildActionNames } from '../actions/utils/asyncUtils';

// pokemon
export const GET_POKEMON = buildActionNames('GET_POKEMON');
export const GET_POKEMON_DETAILS = buildActionNames('GET_POKEMON_DETAILS');
