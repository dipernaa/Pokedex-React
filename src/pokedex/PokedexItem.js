import React from 'react';
import { Card, CardHeader, CardText } from 'material-ui/Card';

const PokedexItem = ({ getPokemonDetail, pokemon }) => (
  <Card
    onExpandChange={() => getPokemonDetail(pokemon.url)}
  >
    <CardHeader
      title={pokemon.name}
      actAsExpander
      showExpandableButton
    />
    <CardText expandable>
      hi
    </CardText>
  </Card>
);

export default PokedexItem;
