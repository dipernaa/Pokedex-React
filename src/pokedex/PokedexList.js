import React, { Component } from 'react';
import { browserHistory } from 'react-router';
import { List, ListItem } from 'material-ui/List';
import TextField from 'material-ui/TextField';
import { routeNames } from '../routes';

const styles = {
  container: {
    display: 'flex',
    flexDirection: 'column'
  },
  search: {
    marginLeft: 'auto',
    marginRight: 'auto'
  }
};

class PokedexList extends Component {

  constructor(props) {
    super(props);

    this.state = {
      filteredPokedex: props.pokedex
    };
  }

  filterPokedex = (value) => {
    if (value) {
      this.setState({
        filteredPokedex: this.props.pokedex.filter(item => item.name.indexOf(value) >= 0)
      });
    } else {
      this.setState({
        filteredPokedex: this.props.pokedex
      });
    }
  }

  render() {
    if (!this.state.filteredPokedex) {
      return null;
    }

    return (
      <div style={styles.container}>
        <TextField
          hintText="Search"
          onChange={(event, value) => this.filterPokedex(value)}
          style={styles.search}
        />
        <br/>

        <div>
          <List>
            {this.state.filteredPokedex.map((pokemon, index) => (
              <ListItem
                key={index}
                className="capitalize"
                onClick={() => browserHistory.push(`${routeNames.detail}?url=${pokemon.url}`)}
                primaryText={pokemon.name}
              />
            ))}
          </List>
        </div>
      </div>
    );
  }
}

export default PokedexList;
