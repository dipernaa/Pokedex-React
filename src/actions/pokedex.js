import api from './utils/api';
import { createAsyncAction } from './utils/asyncUtils';
import { GET_POKEMON } from '../constants/actionTypes';

function getPokemon() {
  return createAsyncAction(GET_POKEMON, () => api.getAsync('/api/pokedex'));
}

export default {
  getPokemon
};
