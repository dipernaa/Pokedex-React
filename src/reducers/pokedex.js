import { GET_POKEMON } from '../constants/actionTypes';

const sortPokemon = (arr) => (
  arr.sort((first, second) => {
    if (first.name > second.name) {
      return 1;
    } else if (first.name < second.name) {
      return -1;
    }

    return 0;
  })
);

export const initialState = {
  count: 0,
  next: null,
  previous: null,
  results: [],
  isLoading: false
};

export default function(state = initialState, action) {
  switch (action.type) {
    case GET_POKEMON.request:
      return {
        ...state,
        isLoading: true
      };

    case GET_POKEMON.failure:
      return {
        ...state,
        isLoading: false
      };

    case GET_POKEMON.success:
      return {
        ...state,
        count: action.data.count,
        next: action.data.next,
        previous: action.data.previous,
        results: sortPokemon(action.data.results.map((item, index) => ({
          ...item,
          id: index + 1
        }))),
        isLoading: false
      };

    default:
      return state;
  }
}
