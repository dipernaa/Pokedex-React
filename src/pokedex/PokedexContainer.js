import React, { Component } from 'react';
import { connect } from 'react-redux';
import muiThemeable from 'material-ui/styles/muiThemeable';
import PokedexItem from './PokedexItem';
import Loading from '../shared/Loading';
import { getPokemon, getPokemonDetails } from '../actions';

export class PokedexContainer extends Component {

  componentWillMount() {
    this.props.getPokemon();
  }

  render() {
    const { getPokemonDetail, isLoading, pokemonData } = this.props;

    if (isLoading) {
      return <Loading />;
    }

    return (
      <div>
        {pokemonData.map((pokemon, index) => (
          <PokedexItem
            key={index}
            isLoading={isLoading}
            getPokemonDetail={getPokemonDetail}
            pokemon={pokemon}
          />
        ))}
      </div>
    );
  }
}

export const mapStateToProps = (state) => {
  const { pokemon } = state;

  return {
    isLoading: pokemon.isLoading,
    pokemonData: pokemon.results
  };
};

export const mapDispatchToProps = (dispatch) => ({
  getPokemon: () => dispatch(getPokemon()),
  getPokemonDetail: (url) => dispatch(getPokemonDetails(url))
});

export default muiThemeable()(connect(mapStateToProps, mapDispatchToProps)(PokedexContainer));
