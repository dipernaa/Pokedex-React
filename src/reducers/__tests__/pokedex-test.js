import pokedex, { initialState } from '../pokedex';
import { GET_POKEMON } from '../../constants/actionTypes';

it('handles unexpected action', () => {
  expect(pokedex(initialState, { type: '_NULL' }))
    .toMatchSnapshot();
});

it('handles GET_POKEMON_REQUEST action', () => {
  expect(pokedex(initialState, { type: GET_POKEMON.request }))
    .toMatchSnapshot();
});

it('handles GET_POKEMON_SUCCESS action', () => {
  const pokemonList = {
    count: 3,
    next: 'http://www.some.url',
    previous: 'http://www.other.url',
    results: [
      { name: 'pokedude' },
      { name: 'ultraPoke' },
      { name: 'slappyPoke' }
    ]
  };

  expect(pokedex(initialState, { type: GET_POKEMON.success, data: pokemonList }))
    .toMatchSnapshot();
});

it('handles GET_POKEMON_FAILURE action', () => {
  expect(pokedex(initialState, { type: GET_POKEMON.failure }))
    .toMatchSnapshot();
});
