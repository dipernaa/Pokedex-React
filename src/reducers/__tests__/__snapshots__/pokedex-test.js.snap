// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles GET_POKEMON_FAILURE action 1`] = `
Object {
  "count": 0,
  "isLoading": false,
  "next": null,
  "previous": null,
  "results": Array [],
}
`;

exports[`handles GET_POKEMON_REQUEST action 1`] = `
Object {
  "count": 0,
  "isLoading": true,
  "next": null,
  "previous": null,
  "results": Array [],
}
`;

exports[`handles GET_POKEMON_SUCCESS action 1`] = `
Object {
  "count": 3,
  "isLoading": false,
  "next": "http://www.some.url",
  "previous": "http://www.other.url",
  "results": Array [
    Object {
      "id": 1,
      "name": "pokedude",
    },
    Object {
      "id": 3,
      "name": "slappyPoke",
    },
    Object {
      "id": 2,
      "name": "ultraPoke",
    },
  ],
}
`;

exports[`handles unexpected action 1`] = `
Object {
  "count": 0,
  "isLoading": false,
  "next": null,
  "previous": null,
  "results": Array [],
}
`;
